### PR TITLE
MathExtras: template'ize alignToPowerOf2

### PR DESCRIPTION
**Pull request by artagnon**
Follow up on 5627794 (MathExtras: avoid unnecessarily widening types) to change the overflow behavior of alignToPowerOf2 to only overflow if the result is not representable in the return type. This allows us to template'ize it, and avoid unnecessarily widening the types of arguments.

### CODE REVIEW

File: llvm/include/llvm/Support/MathExtras.h [hunk 1]
```diff
@@ -488,13 +488,21 @@ constexpr uint64_t alignTo(uint64_t Value, uint64_t Align) {
...
+}
+
+/// Fallback when arguments aren't integral.
 constexpr uint64_t alignToPowerOf2(uint64_t Value, uint64_t Align) {
   assert(Align != 0 && (Align & (Align - 1)) == 0 &&
          "Align must be a power of 2");
-  // Replace unary minus to avoid compilation error on Windows:
-  // "unary minus operator applied to unsigned type, result still unsigned"
-  uint64_t NegAlign = (~Align) + 1;
-  return (Value + Align - 1) & NegAlign;
```
Comment (jayfoad) [inline llvm/include/llvm/Support/MathExtras.h:497]: What was wrong with the old code?

Comment (jayfoad) [reply llvm/include/llvm/Support/MathExtras.h:497]: I think: `Align - 1` cannot overflow and `Value + (Align - 1)` can only overflow if the result is going to overflow to 0 anyway. Unless you have a counterexample?

Comment (jayfoad) [reply llvm/include/llvm/Support/MathExtras.h:497]: If Align is 1 it does not and should not overflow. If Align is 2 it should overflow to 0. I still don't see what is wrong with the old code.

Comment (artagnon) [reply llvm/include/llvm/Support/MathExtras.h:497]: Sorry about taking so long to debug this, but I finally have definitive proof:

Consider Value=8589934592, Align=1, U = unsigned int, V = unsigned long. Then,

T is computed as unsigned long.
NegAlign is computed as 4294967295 (unsigned long).
Value + (Align - 1) is computed as 8589934592 (unsigned long).
(Value + (Align - 1)) & NegAlign is computed as 0 (unsigned long).

**Comment by artagnon** [comment]
Gentle ping. @jayfoad: What do you think of the counter-example?

File: llvm/include/llvm/Support/MathExtras.h (hunk 1 above)
Comment (jayfoad) [reply llvm/include/llvm/Support/MathExtras.h:497]: I assume this is with long being 64 bits and int being 32 bits? This is just a bug in the way you're calculating NegAlign. You need to extend Align to type T before negating, e.g. with `T NegAlign = (T)0 - Align;`

Comment (artagnon) [reply llvm/include/llvm/Support/MathExtras.h:497]: Thanks, fixed now.

File: llvm/include/llvm/Support/MathExtras.h [hunk 2]
```diff
@@ -488,13 +488,21 @@ constexpr uint64_t alignTo(uint64_t Value, uint64_t Align) {
...
+  uint64_t NegAlign = static_cast<uint64_t>(0) - Align;
```
Comment (jayfoad) [inline llvm/include/llvm/Support/MathExtras.h:504]: Nit: don't need any casts here since everything is uint64_t already.

**Review by jayfoad — APPROVED** [review]
LGTM, thanks.

**Pull request merged**